Update Helix.md to include process for updating helix matrix (#29434)
dotnet/aspnetcore: docs/Helix.md

- Skipping an individual test via `[SkipOnHelix("url to github issue")]`.

Make sure to file an issue for any skipped tests and include that in a comment next to either of these

## Process for updating helix matrix

Goal is to balance cost/flakiness against having some coverage of supported distros:
- At the start of each product version, we pick a set of queues/versions/arches to run based on popularity and perceived risk, and how long is left in the support for that OS version.

> - Whenever a new OS is coming online, we ask CTI to do a run on it, and we submit a PR to add a queue for it to check for any failures in it, but if there aren’t any, we don’t merge it.

Reviewer: - While I understand the need for a Helix queue for non-CTI testing, this reads like nobody else would create one for the new OS version. Suggest we call that out as optional.
    - No need to add a Helix queue if we _only_ use CTI tests to confirm the new OS version is fine.
    - When we need a Helix queue, it would be because we plan to at least open a dotnet/aspnetcore PR to try the queue out. Wouldn't the Helix queue update (from us or whoever) need to be merged for a PR in our repo❔
    - More generally, isn't it CTI testing "and / or" try out a dotnet/aspnetcore PR using a Helix queue for the new OS version❔

Author: Sure did you have a particular wording you had in mind?  The gist of it sounds like we will look at updating the queues at the start of the product cycle, and mostly use CTI to validate new asks during each cycle.  I think the main value of this blurb is mostly to capture intent, as opposed to calling out the specifics of how we are going to do each of these steps, since we should just 'do the right thing'

Author: But yeah for #2, there's no possibility of exercising a new os distro without the helix docker image in place, which if we are being proactive and doing this at the start of release cycles, will probably be us that needs to spin them up...

> - Whenever a new OS is coming online, we ask CTI to do a run on it, and we submit a PR to add a queue for it to check for any failures in it, but if there aren’t any, we don’t merge it.

Reviewer 2: ```suggestion
- Whenever a new OS is coming online, we ask CTI to do a run on it, and if there is support for it in helix, we submit a PR to update our helix-matrix to include it for it to check for any failures in it, but if there aren’t any, we don’t merge it.
```

> - Whenever a new OS is coming online, we ask CTI to do a run on it, and if there is support for it in helix, we submit a PR to update our helix-matrix to include it for it to check for any failures in it, but if there aren’t any, we don’t merge it.

Reviewer: Suggest mentioning potential need to update dotnet/dotnet-buildtools-prereqs-docker if Helix support doesn't (yet) exist e.g.
```suggestion
- Whenever a new OS is coming online, we ask CTI to do a run on it, and if there is support for it in helix, we submit a PR to update our helix-matrix to include it for it to check for any failures in it, but if there aren’t any, we don’t merge it.
    - If an appropriate queue does not yet exist, we could submit a PR to https://github.com/dotnet/dotnet-buildtools-prereqs-docker to add it. This helps even if we do not plan to keep the dotnet/aspnetcore change around.
```

> - Link to [OS support calendar] (https://dev.azure.com/devdiv/DevDiv/_wiki/wikis/DevDiv.wiki/12624/OS-Version-Management-Calendar-2021)

Reviewer: Link syntax is invalid with a space there...
```suggestion
- Link to [OS support calendar](https://dev.azure.com/devdiv/DevDiv/_wiki/wikis/DevDiv.wiki/12624/OS-Version-Management-Calendar-2021)
```

- Link to [current list of queues](../eng/targets/Helix.Common.props)

## Example of adding a new docker image to helix

> - Example PR: https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/398

Reviewer: Do you also have an example dotnet/aspnetcore PR using the queue added in dotnet/dotnet-buildtools-prereqs-docker#398❔

nit (not sure why GitHub doesn't recognize the URI):
```suggestion
- Example PR: dotnet/dotnet-buildtools-prereqs-docker#398
```

Author: That PR after will come once the queue is up and healthy, as we'll need to switch fedora over to it, I can link that PR to the docs as part of the PR

## Process for updating helix matrix

Goal is to balance cost/flakiness against having some coverage of supported distros:
- At the start of each product version, we pick a set of queues/versions/arches to run based on popularity and perceived risk, and how long is left in the support for that OS version.
- Whenever a new OS is coming online, we ask CTI to do a run on it, and if there is support for it in helix, we submit a PR to update our helix-matrix to include it for it to check for any failures in it, but if there aren’t any, we don’t merge it.
- Link to [OS support calendar](https://dev.azure.com/devdiv/DevDiv/_wiki/wikis/DevDiv.wiki/12624/OS-Version-Management-Calendar-2021)
- Link to [current list of queues](../eng/targets/Helix.Common.props)

## Example of adding a new docker image to helix

- Example PR: dotnet/dotnet-buildtools-prereqs-docker#398
- Summary is to update [manifest.json](https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/master/manifest.json) with an entry for the new dockerfiles, and then add the docker files as well to dotnet-buildtools-prereqs-docker

> - The resulting new docker queue id will be found in: https://github.com/dotnet/versions/blob/master/build-info/docker/image-info.dotnet-dotnet-buildtools-prereqs-docker-master.json

Reviewer: nit #2:
```suggestion
- The resulting new docker queue id will be found in: [image-info.dotnet-dotnet-buildtools-prereqs-docker-master.json](https://github.com/dotnet/versions/blob/master/build-info/docker/image-info.dotnet-dotnet-buildtools-prereqs-docker-master.json)
```
